### PR TITLE
refactor: score calculation naming

### LINKED ISSUE #1137: refactor: fix searching via bm25 in playground by enhancing `semantic_scores` to `score_calculation` to simplify playground and BFF
The issue is that while the indexer receives `semantic_scores` with `bm25` it doesn't receive `apply_default_bm25_query` as the BFF never exposes this parameter. Proof of that is given in second message.

The better way is to get rid off extra handling of scores and change the interface to also incorporate bm25 in playground and BFF by default. Concretely

- [ ] remove `apply_default_bm25`
- [ ] remove `custom_bm25_query`
- [ ] remove default `bm25` field in elastic indexer
- [ ] enhance `semantic_scores` to be of format `score_calculation`: `(query_field, index_field, matching_method, weight)`:
  - `matching_method` describes how the `query_field` and `index_field` should be matched (semantic search via CLIP or S-BERT, bm25, etc.); it's a string to indicate `encoderclip` or `encodersbert` as before or `bm25` to indicate search by bm25
  - `score_calculation` gives an overall intuitive description on how to match a query
- [ ] refactor playground to match new format of `score_calculation` by 
  - removing extra handling of bm25

### PR DESCRIPTION
**Pull request by joschkabraun**
Closes #1137 and:

- fixes playground to enable bm25 search again
- renames `semantic_scores`  to more interpretable `score_calculation`
- elastic indexer doesn't use `custom_bm25_query` or `apply_default_bm25` anymore but one can simply specify those via specifying `matching_method` in `score_calculation` to `bm25` (`matching_method` can be an encoder name or `bm25` now)
- simplifies user interface in playground by removing extra `bm25` button and setting it with the score calculation

---

- [x] This PR references an open issue
- [x] Added a test
- [x] Updated the documentation
- [x] Added a screenshot of a successful customer deployment

Regular search: 

Screenshot 2023-03-03 at 15 27 58

BM25 search:

Screenshot 2023-03-03 at 15 28 20

### CODE REVIEW

File: docs/user-guides/cli_api.md [hunk 1]
```diff
@@ -112,7 +112,7 @@ Optional parameters:
     - example: `5`
   - **filters**: a dictionary of filters, with filter name as key and target as value
     - example: `{'tags__color': {'$eq': 'blue'}, 'tags__price': {'$gt': 100}}`
-  - **semantic_scores**: list of semantic scores defining how fields should be compared and weighted. Each semantic score
+  - **score_calculation**: list of score calculation defining how fields should be compared and weighted. Each score calculation
```
Comment (violenil) [inline docs/user-guides/cli_api.md:115]:
```suggestion
  - **score_calculation**: list of score calculation components defining how fields should be compared and weighted. Each score calculation
```

File: now/executor/gateway/playground/playground.py [hunk 2]
```diff
@@ -340,16 +340,9 @@ class UI:
     )
 
 
-def delete_semantic_scores():
-    st.session_state['len_semantic_scores'] = 0
-    st.session_state.semantic_scores = {}
-
-
-def toggle_bm25_slider():
```
Comment (violenil) [inline now/executor/gateway/playground/playground.py:348]: Remind me what we are doing with bm25 in the playground, are we removing it all-together?

File: now/executor/indexer/elastic/elastic_indexer.py [hunk 3]
```diff
@@ -215,10 +218,9 @@ def search(
                 - 'limit' (int): Number of matches to get per Document, default 100.
                 - 'get_score_breakdown' (bool): Wether to return the score breakdown, i.e. the scores of each
                     field+encoder combination/comparison.
-                - 'apply_default_bm25' (bool): Whether to apply the default bm25 scoring. Default is False. Will
-                    be ignored if 'custom_bm25_query' is specified.
-                - 'custom_bm25_query' (dict): Custom query to use for BM25. Note: this query can only be
-                    passed if also passing `es_mapping`. Otherwise, only default bm25 scoring is enabled.
+                - 'score_calculation' (List[Tuple]): list of tuples of (query_field, document_field, matching_method,
```
Comment (violenil) [inline now/executor/indexer/elastic/elastic_indexer.py:221]: The score_calculation that arrives at the indexer is actually a `List[List]` and not `List[Tuple]`, we cannot pass tuples through the flow. If they are sent to bff as list of tuples, the bff converts them to list.

Comment (violenil) [reply now/executor/indexer/elastic/elastic_indexer.py:221]: Perhaps we should update the `generate_score_calculation` to behave the same way and generate the same types.

**Review by violenil — COMMENTED** [review]
Looks great! I just have a few questions, see comments. Does bm25 work in the api? And are we removing bm25 entirely from the playground or did I miss something?

File: now/executor/gateway/playground/playground.py (hunk 2 above)
Comment (joschkabraun) [reply now/executor/gateway/playground/playground.py:348]: No, we can select it in the score calculation dropdown as shown in the screenshot.

File: now/executor/indexer/elastic/elastic_indexer.py (hunk 3 above)
Comment (joschkabraun) [reply now/executor/indexer/elastic/elastic_indexer.py:221]: Good point!

**Review by jupyterjazz — APPROVED** [review]
Nice refactoring!

File: now/executor/gateway/bff/app/v1/models/search.py [hunk 4]
```diff
@@ -49,10 +49,11 @@ class SearchRequestModel(BaseRequestModel):
         'This is useful if the file is stored in a cloud bucket.',
         example=False,
     )
-    semantic_scores: List[Tuple] = Field(
+    score_calculation: List[Tuple] = Field(
```
Comment (violenil) [inline now/executor/gateway/bff/app/v1/models/search.py:52]: Do you mind also updating the types here? And in the example below, to show that the `score_calculation` is actually expected to be a `List[List]`

File: now/executor/gateway/bff/app/v1/routers/search.py [hunk 5]
```diff
@@ -187,35 +187,30 @@ async def search(
...
+    :return: the score calculation in the format expected by the indexer. Example:
+        [['query_text', 'my_product_image', 'encoderclip', 1], ['query_text', 'my_product_text', 'bm25', 1]]
     """
-    semantic_scores = []
-    for sem_score in data.semantic_scores:
-        sem_score = list(sem_score)
-        sem_score[0] = field_names_to_dataclass_fields[sem_score[0]]
+    score_calculation = []
+    for scr_calc in data.score_calculation:
+        scr_calc = list(scr_calc)
```
Comment (violenil) [inline now/executor/gateway/bff/app/v1/routers/search.py:204]: If we always show that the `score_calculation` is now expected to be a `List[List]` then we probably also don't need this casting anymore (just need to verify that the tests all work 😄 )

File: now/executor/indexer/elastic/es_query_building.py [hunk 6]
```diff
@@ -49,15 +49,13 @@ def generate_semantic_scores(
                     )
```
Comment (violenil) [inline now/executor/indexer/elastic/es_query_building.py:49]: Should be a list, right?

File: tests/conftest.py [hunk 7]
```diff
@@ -218,10 +232,10 @@ class MMQuery:
 
     document_mappings = [['clip', 8, ['title', 'gif']]]
 
-    default_semantic_scores = [
+    default_score_calculation = [
         ('query_text', 'title', 'clip', 1),
```
Comment (violenil) [inline tests/conftest.py:236]: For consistency, these should be lists too.

File: now/executor/gateway/bff/app/v1/models/search.py [hunk 8]
```diff
@@ -49,10 +49,11 @@ class SearchRequestModel(BaseRequestModel):
         'This is useful if the file is stored in a cloud bucket.',
         example=False,
     )
-    semantic_scores: List[Tuple] = Field(
+    score_calculation: List[List] = Field(
         default=[],
-        description='List of tuples where each tuple contains a query_field, index_field, encoder_name and weight.'
-        ' This defines how scores should be calculated for documents.',
+        description='List of tuples where each tuple contains a query_field, index_field, matching_method and weight.'
```
Comment (violenil) [inline now/executor/gateway/bff/app/v1/models/search.py:54]:
```suggestion
        description='List of lists, where each nested list contains a query_field, index_field, matching_method and weight.'
```

File: now/executor/gateway/bff/app/v1/models/search.py [hunk 9]
```diff
@@ -49,10 +49,11 @@ class SearchRequestModel(BaseRequestModel):
...
+        ' This defines how scores should be calculated for documents. The matching_method can be an encoder name or '
+        'bm25. The weight is a float which is used to scale the score.',
         example=[('query_text', 'title', 'encoderclip', 1.0)],
```
Comment (violenil) [inline now/executor/gateway/bff/app/v1/models/search.py:57]:
```suggestion
        example=[['query_text', 'title', 'encoderclip', 1.0]],
```

File: now/executor/gateway/bff/app/v1/routers/search.py [hunk 10]
```diff
@@ -72,7 +72,7 @@
                 }
             ],
             'create_temp_link': False,
-            'semantic_scores': [('query_text_0', 'title', 'encoderclip', 1.0)],
+            'score_calculation': [('query_text_0', 'title', 'encoderclip', 1.0)],
```
Comment (violenil) [inline now/executor/gateway/bff/app/v1/routers/search.py:75]:
```suggestion
            'score_calculation': [['query_text_0', 'title', 'encoderclip', 1.0]],
```

File: now/executor/gateway/playground/playground.py [hunk 11]
```diff
@@ -439,7 +400,7 @@ def customize_semantic_scores():
             value=0.5,
             key='weight_' + str(i),
         )
-        st.session_state.semantic_scores[f'{i}'] = (q_field, id_field, enc, w)
+        st.session_state.score_calculation[f'{i}'] = (q_field, id_field, enc, w)
```
Comment (violenil) [inline now/executor/gateway/playground/playground.py:403]:
```suggestion
        st.session_state.score_calculation[f'{i}'] = [q_field, id_field, enc, w]
```
Without this change, and with conversion to list removed in bff, I imaging this should be breaking custom score calculation in the playground?

File: now/executor/indexer/elastic/es_query_building.py [hunk 12]
```diff
@@ -70,14 +68,11 @@ def build_es_queries(
...
     :param docs_map: dictionary mapping encoder to DocumentArray.
-    :param apply_default_bm25: whether to combine bm25 with vector search. If False,
-        will only perform vector search. If True, must supply a text
-        field for bm25 searching.
     :param get_score_breakdown: whether to return the score breakdown for matches.
         For this function, this parameter determines whether to return the embeddings
         of a query document.
-    :param semantic_scores: list of semantic scores used to calculate a score for a document.
-    :param custom_bm25_query: custom query to use for BM25.
+    :param score_calculation: list of tuples of (query_field, document_field, matching_method, linear_weight) to show
```
Comment (violenil) [inline now/executor/indexer/elastic/es_query_building.py:74]:
```suggestion
    :param score_calculation: list of nested lists containing (query_field, document_field, matching_method, linear_weight) which define
```

**Review by violenil — APPROVED** [review]
After some final comments are addressed, this LGTM!🚀

**Pull request merged**